### ISSUE TITLE
About ELQ

### ISSUE THREAD

**Issue by pengyxhack** [HXX97/GMT-KBQA#3]
Hello, I have a problem when I do entity linking with ELQ service. 

`from detect_and_link_entity import get_entity_linking_from_elq`
`mention = "what religions are practiced in the country that has the national anthem Afghan National Anthem"`
`entity = get_entity_linking_from_elq(mention)`
`print(entity)`

I ran the code above, but I only got the output following:

`[{'id': [], 'label': 'Afghan National Anthem', 'mention': 'afghan national anthem', 'score': -1.2437682151794434, 'perfect_match': True}, {'id': [], 'label': 'Religion', 'mention': 'religions', 'score': -3.7547619342803955, 'perfect_match': False}, {'id': [], 'label': 'Islam', 'mention': 'religions', 'score': -4.306562900543213, 'perfect_match': False}, {'id': [], 'label': 'Religion in Afghanistan', 'mention': 'religions', 'score': -5.706789016723633, 'perfect_match': False}, {'id': [], 'label': 'Abrahamic religions', 'mention': 'religions', 'score': -6.387351989746094, 'perfect_match': False}, {'id': [], 'label': 'Religion in Bangladesh', 'mention': 'religions', 'score': -6.408331871032715, 'perfect_match': False}, {'id': [], 'label': 'National anthem of the Kingdom of Afghanistan', 'mention': 'afghan national anthem', 'score': -7.044179439544678, 'perfect_match': False}, {'id': [], 'label': 'National anthem of the Democratic Republic of Afghanistan', 'mention': 'afghan national anthem', 'score': -8.992083549499512, 'perfect_match': False}, {'id': [], 'label': 'National anthem', 'mention': 'afghan national anthem', 'score': -9.269025802612305, 'perfect_match': False}, {'id': [], 'label': 'Anthem of the Republic of Kazakhstan', 'mention': 'afghan national anthem', 'score': -10.114151000976562, 'perfect_match': False}]`
Why can't I get the freebase id?

**Comment by WuXuan374** [comment]
Hello, I first re-run the following code 

    from detect_and_link_entity import get_entity_linking_from_elq

    mention = "what religions are practiced in the country that has the national anthem Afghan National Anthem"

    entity = get_entity_linking_from_elq(mention)

    print(entity)
and the result is

    [{'id': 'm.02r0hl7', 'label': 'Afghan National Anthem', 'mention': 'afghan national anthem', 'score': -1.2437647581100464, 'perfect_match': True}, {'id': 'm.06bvp', 'label': 'Religion', 'mention': 'religions', 'score': -3.75476336479187, 'perfect_match': False}, {'id': 'm.0flw86', 'label': 'Islam', 'mention': 'religions', 'score': -4.306560516357422, 'perfect_match': False}, {'id': 'm.011brgyq', 'label': 'Religion in Afghanistan', 'mention': 'religions', 'score': -5.706789970397949, 'perfect_match': False}, {'id': 'm.0f_h6', 'label': 'Abrahamic religions', 'mention': 'religions', 'score': -6.387356758117676, 'perfect_match': False}, {'id': 'm.0ch3ycm', 'label': 'Religion in Bangladesh', 'mention': 'religions', 'score': -6.408329486846924, 'perfect_match': False}, {'id': 'm.0134vyr3', 'label': 'National anthem of the Kingdom of Afghanistan', 'mention': 'afghan national anthem', 'score': -7.044186592102051, 'perfect_match': False}, {'id': 'm.01j2gd', 'label': 'National anthem', 'mention': 'afghan national anthem', 'score': -9.26904296875, 'perfect_match': False}, {'id': 'm.01qqsx', 'label': 'Anthem of the Republic of Kazakhstan', 'mention': 'afghan national anthem', 'score': -10.114145278930664, 'perfect_match': False}]


Then I check the code of `get_entity_linking_from_elq`, and I find in [detect_and_link_entity.py L209](https://github.com/HXX97/GMT-KBQA/blob/main/detect_and_link_entity.py#L209) there is a SPARQL query to convert wikiIDs to freebase mids. The content of the SPARQL query is in [sparql_executor.py L959](https://github.com/HXX97/GMT-KBQA/blob/main/executor/sparql_executor.py#L959) as follows:

        SPARQL
        PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
        PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
        PREFIX : <http://rdf.freebase.com/ns/> 
        SELECT (?x0 AS ?value) WHERE {
        SELECT DISTINCT ?x0  WHERE {
        """
              '?x0 <http://rdf.freebase.com/key/wikipedia.en_id> ' + f'"{wikiID}"'
                             """
    FILTER regex(?x0, "http://rdf.freebase.com/ns/")
    }
    }

I guess this could be the issue. Perhaps you can check whether this SPARQL query returns expected results. Feel free to contact us, if there are any further questions.

**Comment by pengyxhack** [comment]
Yes, I just can't convert the wikiIDs to freebase mids. I execute the sparql without odbc. The code is as follows:

query2 = ("""
        PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
        PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
        PREFIX : <http://rdf.freebase.com/ns/> 
        SELECT (?x0 AS ?value) WHERE {
        SELECT DISTINCT ?x0  WHERE {
        """
              '?x0 <http://rdf.freebase.com/key/wikipedia.en_id> ' + f'"{wikiID}"'
                             """
    FILTER regex(?x0, "http://rdf.freebase.com/ns/")
    }
    }
    """)
    sparql.setQuery(query2)
    try:
        results = sparql.query().convert()
    except urllib.error.URLError:
        print(query2)
        # exit(0)
    rtn = []
    for result in results['results']['bindings']:
        #assert len(result) == 1  # only select one variable
        for var in result:
            rtn.append(result[var]['value'].replace('http://rdf.freebase.com/ns/', '').replace("-08:00", ''))`

**Comment by WuXuan374** [comment]
> Yes, I just can't convert the wikiIDs to freebase mids. I execute the sparql without odbc. The code is as follows:
> 
> query2 = (""" PREFIX rdf: http://www.w3.org/1999/02/22-rdf-syntax-ns# PREFIX rdfs: http://www.w3.org/2000/01/rdf-schema# PREFIX : http://rdf.freebase.com/ns/ SELECT (?x0 AS ?value) WHERE { SELECT DISTINCT ?x0 WHERE { """ '?x0 http://rdf.freebase.com/key/wikipedia.en_id ' + f'"{wikiID}"' """ FILTER regex(?x0, "http://rdf.freebase.com/ns/") } } """) sparql.setQuery(query2) try: results = sparql.query().convert() except urllib.error.URLError: print(query2) # exit(0) rtn = [] for result in results['results']['bindings']: #assert len(result) == 1 # only select one variable for var in result: rtn.append(result[var]['value'].replace('http://rdf.freebase.com/ns/', '').replace("-08:00", ''))`

Hi, comparing the code, maybe you can try

'?x0 \<http://rdf.freebase.com/key/wikipedia.en_id>' + f'"{wikiID}"'

instead of

'?x0 http://rdf.freebase.com/key/wikipedia.en_id ' + f'"{wikiID}"'

i.e., '<' and '>' around the uri

Query and answers attached below, I execute the query using our Virtuoso UI.
<img width="634" alt="截屏2023-05-31 下午3 34 33" src="https://github.com/HXX97/GMT-KBQA/assets/48612702/c285c140-e594-4a41-a90d-e2b261e12838">
<img width="354" alt="截屏2023-05-31 下午3 34 37" src="https://github.com/HXX97/GMT-KBQA/assets/48612702/2f7af3b5-afe6-4b54-bf9d-94730c22ffd1">

**Comment by pengyxhack** [comment]
Sorry for the typo, but the actual code is what you said, i.e., '<' and '>' around the uri. But I still can't get the freebase mid.

**Comment by WuXuan374** [comment]
It's weird. Sorry, but so far I cannot think of other possible reasons.
